nav2_recoveries/src/recovery_server.cpp

```diff
@@ -0,0 +1,145 @@
+// Copyright (c) 2018 Samsung Research America
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License. Reserved.
+
+#include <memory>
+#include <string>
+#include <vector>
+#include "nav2_recoveries/recovery_server.hpp"
+
+namespace recovery_server
+{
+
+RecoveryServer::RecoveryServer()
+: nav2_util::LifecycleNode("nav2_recoveries", "", true),
+  plugin_loader_("nav2_core", "nav2_core::Recovery")
+{
+}
+
+
+RecoveryServer::~RecoveryServer()
+{
+}
+
+nav2_util::CallbackReturn
+RecoveryServer::on_configure(const rclcpp_lifecycle::State & /*state*/)
+{
+  RCLCPP_INFO(get_logger(), "Configuring");
+
+  tf_ = std::make_shared<tf2_ros::Buffer>(get_clock());
+  auto timer_interface = std::make_shared<tf2_ros::CreateTimerROS>(
+    get_node_base_interface(),
+    get_node_timers_interface());
+  tf_->setCreateTimerInterface(timer_interface);
+  transform_listener_ = std::make_shared<tf2_ros::TransformListener>(*tf_);
+
+  declare_parameter("costmap_topic",
+    rclcpp::ParameterValue(std::string("local_costmap/costmap_raw")));
+  declare_parameter("footprint_topic",
+    rclcpp::ParameterValue(std::string("local_costmap/published_footprint")));
+
+  std::vector<std::string> plugin_names{std::string("Spin"),
+    std::string("BackUp")};
+  std::vector<std::string> plugin_types{std::string("nav2_recoveries/Spin"),
```

Comment:
If we had a long list of plugins, I'd recommend setting a hash table instead of two separte vectors:

```cpp
std::unordered_map<std::string, std::string> plugins_{
    {"Spin", "nav2_recoveries/Spin"},
    ...}
))
```

Reply:
agreed, but not implemented in parameters yet in rclcpp

Reply:
I took that from the costmap layer method when I was surprised to find that the map parameter readin wasn't implemented in ROS2

Reply:
Once that's around I'll update it everywhere because you're right, this is ugly